File: app/src/main/java/br/com/jonathanzanella/myexpenses/account/AccountAdapterPresenter.java
package br.com.jonathanzanella.myexpenses.account;

import android.os.AsyncTask;

import java.util.List;

class AccountAdapterPresenter {
	private final AccountRepository repository;
	private final AccountAdapter adapter;

	private List<Account> accounts;

	AccountAdapterPresenter(AccountAdapter adapter, AccountRepository repository, final AccountAdapter.Format format) {
		this.repository = repository;
		this.adapter = adapter;
	}

	final void loadAccountsAsync(final AccountAdapter.Format format) {
		new AsyncTask<Void, Void, Void>() {

			@Override
			protected Void doInBackground(Void... voids) {
				if(format == AccountAdapter.Format.RESUME) {
					accounts = repository.forResumeScreen();
				} else {
					accounts = repository.all();
				}
				return null;
			}

			@Override
			protected void onPostExecute(Void aVoid) {
				super.onPostExecute(aVoid);
				adapter.notifyDataSetChanged();
			}
		}.execute();
	}

	Account getAccount(int position) {
		return accounts.get(position);
	}

	int getAccountsSize() {
		return accounts == null ? 0 : accounts.size();
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/account/AccountAdapterPresenter.kt
package br.com.jonathanzanella.myexpenses.account

import br.com.jonathanzanella.myexpenses.App
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import javax.inject.Inject

class AccountAdapterPresenter(private val adapter: AccountAdapter) {

    private var accounts: List<Account>? = null

    @Inject
    lateinit var repository: AccountRepository

    init {
        App.getAppComponent().inject(this@AccountAdapterPresenter)
    }

    fun loadAccountsAsync(format: AccountAdapter.Format) {
        doAsync {
            accounts = when {
                format === AccountAdapter.Format.RESUME -> repository.forResumeScreen()
                else -> repository.all()
            }

            uiThread { adapter.notifyDataSetChanged() }
        }
    }

    fun getAccount(position: Int): Account {
        return accounts!![position]
    }

    val accountsSize: Int
        get() {
            return accounts?.size ?: 0
        }
}
